src/languageSupport/languages/flux/lsp/connection.ts

```diff
@@ -91,6 +109,203 @@ class LspConnectionManager {
+    const startLine = COMPOSITION_INIT_LINE
+    const endLine =
+      (query.split('\n').findIndex(line => line.includes(COMPOSITION_YIELD)) ||
+        0) + 1
+    return {startLine, endLine}
+  }
+
+  _setSessionSync(synced: boolean) {
+    this._callbackSetSession({
+      composition: {synced},
+    })
+  }
+
+  _setEditorSyncToggle() {
+    setTimeout(() => {
+      // elements in monaco-editor. positioned by editor.
+      const syncIcons = document.getElementsByClassName(ICON_SYNC_CLASSNAME)
+
+      // UI elements we control
+      const clickableInvisibleDiv = document.getElementById(ICON_SYNC_ID)
+      if (!syncIcons.length || !clickableInvisibleDiv) {
+        return
+      }
+
+      const [upperIcon] = syncIcons
+      let [, lowerIcon] = syncIcons
+      if (!lowerIcon) {
+        lowerIcon = upperIcon
+      }
+      const {startLine, endLine} = this._getCompositionBlockLines()
+
+      // move div to match monaco-editor coordinates
+      clickableInvisibleDiv.style.top =
+        ((upperIcon as any).offsetTop || 0) + 'px'
+      const height =
+        ((lowerIcon as any).offsetHeight || 0) * (endLine - startLine + 1) +
+        ((upperIcon as any).offsetTop || 0)
+      clickableInvisibleDiv.style.height = height + 'px'
+      clickableInvisibleDiv.style.width =
+        ((upperIcon as any).offsetWidth || 0) + 'px'
+
+      // add listeners
+      clickableInvisibleDiv.removeEventListener('click', () =>
+        this._setSessionSync(!this._session.composition.synced)
+      ) // may have existing
+      clickableInvisibleDiv.addEventListener('click', () =>
+        this._setSessionSync(!this._session.composition.synced)
+      )
+    }, 1000)
+  }
+
+  _editorChangeIsFromLsp(change) {
+    return !!change.forceMoveMarkers
+  }
+
+  _setEditorIrreversibleExit() {
+    this._model.onDidChangeContent(e => {
+      const {changes} = e
+      changes.some(change => {
+        const {startLine, endLine} = this._getCompositionBlockLines()
+        if (
+          change.range.startLineNumber >= startLine &&
+          change.range.endLineNumber <= endLine &&
+          !this._editorChangeIsFromLsp(change)
+        ) {
+          this._callbackSetSession({
+            composition: {synced: false, diverged: true},
+          })
+          return
+        }
+      })
+    })
+  }
+
+  _setEditorBlockStyle() {
+    const {startLine, endLine} = this._getCompositionBlockLines()
+
+    const startLineStyle = [
+      {
+        range: new MonacoTypes.Range(startLine, 1, startLine, 1),
+        options: {
+          linesDecorationsClassName: ICON_SYNC_CLASSNAME,
+        },
+      },
+    ]
+    const endLineStyle = [
+      {
+        range: new MonacoTypes.Range(endLine, 1, endLine, 1),
+        options: {
+          linesDecorationsClassName: ICON_SYNC_CLASSNAME,
+        },
+      },
+    ]
+
+    this._compositionStyle = this._editor.deltaDecorations(
+      this._compositionStyle,
+      startLineStyle.concat(endLineStyle)
+    )
```

Comment:
I have a whole writeup on why `editor.deltaDecorations()` was chosen for composition block styling. Please feel to ask for details.

FyI: assessed, tested, and rejected options were:
* LSP `textDocument/documentColor` and `textDocument/colorPresentation`
* LSP `textDocument/documentHighlight`
* `monaco.editor.colorizeModelLine()`
* `monaco.editor.colorizeElement()`
* `monaco.editor.colorize()`

Reply:
Thanks for digging into the Monaco Editor APIs!